src/browser/fake_network.js

```diff
@@ -43,6 +84,179 @@ function iptolong(parts) {
+        }
+        const buffer = this.buffer;
+
+        const new_head = this.head + src_length;
+        if(new_head > capacity) {
+            const i_split = capacity - this.head;
+            buffer.set(src_array.subarray(0, i_split), this.head);
+            buffer.set(src_array.subarray(i_split));
+        }
+        else {
+            buffer.set(src_array, this.head);
+        }
+        this.head = new_head % capacity;
+        this.length += src_length;
+    }
+
+    /**
+     * @param {Uint8Array} dst_array
+     */
+    peek(dst_array)
+    {
+        const length = Math.min(this.length, dst_array.length);
+        if(length) {
+            const buffer = this.buffer;
+            const capacity = buffer.length;
+            const new_tail = this.tail + length;
+            if(new_tail > capacity) {
+                const buf_len_left = new_tail % capacity;
+                const buf_len_right = capacity - this.tail;
+                dst_array.set(buffer.subarray(this.tail));
+                dst_array.set(buffer.subarray(0, buf_len_left), buf_len_right);
+            }
+            else {
+                dst_array.set(buffer.subarray(this.tail, new_tail));
+            }
+        }
+        return length;
+    }
+
+    /**
+     * @param {number} length
+     */
+    remove(length)
+    {
+        if(length > this.length) {
+            length = this.length;
+        }
+        if(length) {
+            this.tail = (this.tail + length) % this.buffer.length;
+            this.length -= length;
+        }
+        return length;
+    }
+}
+
+function create_eth_encoder_buf()
+{
+    const eth_frame = new Uint8Array(ETH_FRAME_SIZE);
+    const buffer = eth_frame.buffer;
+    const offset = eth_frame.byteOffset;
+    return {
+        eth_frame: eth_frame,
+        eth_frame_view: new DataView(buffer),
+        eth_payload_view: new DataView(buffer, offset + ETH_PAYLOAD_OFFSET, ETH_PAYLOAD_SIZE),
+        ipv4_payload_view: new DataView(buffer, offset + IPV4_PAYLOAD_OFFSET, IPV4_PAYLOAD_SIZE),
+        udp_payload_view: new DataView(buffer, offset + UDP_PAYLOAD_OFFSET, UDP_PAYLOAD_SIZE),
+        text_encoder: new TextEncoder()
+    };
+}
+
+/**
+ * Copy given data array into view starting at offset, return number of bytes written.
+ *
+ * @param {number} offset
+ * @param {ArrayBuffer|ArrayBufferView} data
+ * @param {DataView} view
+ * @param {Object} out
+ */
+function view_setArray(offset, data, view, out)
+{
+    out.eth_frame.set(data, view.byteOffset + offset);
+    return data.length;
+}
+
+/**
+ * UTF8-encode given string into view starting at offset, return number of bytes written.
+ *
+ * @param {number} offset
+ * @param {string} str
+ * @param {DataView} view
+ * @param {Object} out
+ */
+function view_setString(offset, str, view, out)
+{
+    const ofs = view.byteOffset + offset;
+    const result = out.text_encoder.encodeInto(str, ofs ? out.eth_frame.subarray(ofs) : out.eth_frame);
```

Comment:
Considering that `ofs` is almost never `0`, I'd remove the ternary.

Reply:
Done.